internal/users/localentries/localgroups.go

```diff
@@ -21,85 +28,188 @@ import (
+func (l *lockedGroups) GetCurrentGroups() (groups []types.GroupEntry) {
+	unlock := l.mustRLock()
+	defer unlock()
+
+	return types.DeepCopyGroupEntries(l.currentGroups)
+}
+
+// SaveGroups saves the passed user groups to the local groups file.
+func (l *lockedGroups) SaveGroups(groups []types.GroupEntry) (err error) {
+	defer decorate.OnError(&err, "could not save groups")
+
+	unlock := l.mustLock()
+	defer unlock()
+
+	return l.saveLocalGroups(groups)
+}
+
+// Update synchronizes for the given user the local group list with the current group list from UserInfo.
+func (l *lockedGroups) Update(username string, newGroups []string, oldGroups []string) (err error) {
+	log.Debugf(context.TODO(), "Updating local groups for user %q, new groups: %v, old groups: %v", username, newGroups, oldGroups)
+	defer decorate.OnError(&err, "could not update local groups for user %q", username)
+
+	unlock := l.mustLock()
+	defer unlock()
+
+	allGroups := types.DeepCopyGroupEntries(l.currentGroups)
+	userGroups := l.userLocalGroups(username)
+	currentGroupsNames := sliceutils.Map(userGroups, func(g types.GroupEntry) string {
+		return g.Name
+	})
+
+	groupsToAdd := sliceutils.Difference(newGroups, currentGroupsNames)
+	log.Debugf(context.TODO(), "Adding %q to local groups: %v", username, groupsToAdd)
 	groupsToRemove := sliceutils.Difference(oldGroups, newGroups)
 	// Only remove user from groups which they are part of
-	groupsToRemove = sliceutils.Intersection(groupsToRemove, currentGroups)
-	log.Debugf(context.TODO(), "Removing from local groups: %v", groupsToRemove)
+	groupsToRemove = sliceutils.Intersection(groupsToRemove, currentGroupsNames)
+	log.Debugf(context.TODO(), "Removing %q from local groups: %v", username, groupsToRemove)
 
-	// Do all this in a goroutine as we don't want to hang.
-	for _, g := range groupsToRemove {
-		args := opts.gpasswdCmd[1:]
-		args = append(args, "--delete", username, g)
-		if err := runGPasswd(opts.gpasswdCmd[0], args...); err != nil {
-			return err
-		}
+	if len(groupsToRemove) == 0 && len(groupsToAdd) == 0 {
+		return nil
 	}
-	for _, g := range groupsToAdd {
-		args := opts.gpasswdCmd[1:]
-		args = append(args, "--add", username, g)
-		if err := runGPasswd(opts.gpasswdCmd[0], args...); err != nil {
-			return err
+
+	getCurrentGroupByName := func(name string) *types.GroupEntry {
+		idx := slices.IndexFunc(allGroups, func(g types.GroupEntry) bool { return g.Name == name })
+		if idx == -1 {
+			return nil
 		}
+		return &allGroups[idx]
 	}
 
-	return nil
-}
-
-// getPasswdUsernames gets the passwd entries and returns their usernames.
-func getPasswdUsernames() ([]string, error) {
-	var usernames []string
-	entries, err := GetPasswdEntries()
-	if err != nil {
-		return nil, err
+	usersCompareFunc := func(u string) bool { return u == username }
+	for _, g := range groupsToRemove {
+		group := getCurrentGroupByName(g)
+		if group == nil {
+			continue
+		}
+		group.Users = slices.DeleteFunc(group.Users, usersCompareFunc)
 	}
```

Comment:
```suggestion
	for _, groupName := range groupsToRemove {
		if group := getGroupByName(groupName); group != nil {
			group.Users = slices.DeleteFunc(group.Users, func(u string) bool { return u == username })
		}
	}
```

Reply:
Mh I find it less readable, and mostly we're leaving the happy path

Reply:
> Mh I find it less readable

Because the call to `getGroupByName` and the `group != nil` check is in the same line? That's the idiomatic way to do it in Go. My preference is to do it when (1.) the line doesn't get too long by it, and (2.) the check is exactly what you would expect to be checked - in this case, if I read `if group := ...`, I would expect the check to be `group != nil`. 

What do you mean with "we're leaving the happy path"?

Reply:
I wouldn't apply that rule if the remainder of the happy path is a single line and the unhappy path is empty

Reply:
> I wouldn't apply that rule if the remainder of the happy path is a single line and the unhappy path is empty

That's debatable, though... I prefer having the expected behavior aligned "as left as possible", tbh. This way, we can easily identify cases where things might have gone out of hand (i.e., the conditions), even if it turns out to have more lines of code in the end.